[docs] Introduce deps and the DuckDB resource in the DuckDB guide/tutorial (#15531)
dagster-io/dagster: docs/content/integrations/duckdb/using-duckdb-with-dagster.mdx

# Using Dagster with DuckDB

This tutorial focuses on how to create and interact with DuckDB tables in Dagster's [Software-defined Assets (SDAs)](/concepts/assets/software-defined-assets).

> The `dagster-duckdb` library provides two ways to interact with DuckDB tables. The [resource](TODO-LINK) allows you to directly run SQL queries against tables within an asset's compute function. The [I/O manager](TODO-LINK) transfers the responsibility of storing and loading DataFrames as DuckDB tables to Dagster.

Author: TODO - decide if the TODO links in this paragraph should link to the API docs for the resource and IO manager, or to the concept page for resources and io managers

Reviewer: I feel like concept pages might be better.

---

> This tutorial is divided in two parts. Both parts will create the same assets, but how the data is stored in DuckDb will differ. In [Part 1](#part-1-using-the-duckdb-resource) you will use the DuckDB resource to manually execute SQL queries to create tables and interact with those tables. In [Part 2](#part-2-using-the-duckdb-io-manager) you will transfer the responsibility for generating the SQL to create and interact with tables to the DuckDB I/O manager. When writing your own assets, you may choose one or the other (or both) approaches depending on your storage requirements. See [When to use I/O managers](TODO-LINK) to learn more.

Reviewer: A few suggestions for this paragraph:

1. It should be at the end of the intro, just before the **Prerequisites**. It's in a weird no-man's land right now
2. Consider breaking this up into a list to make it easier to read - ex:

> This tutorial is divided into two parts:
> - In Part 1...
> - In Part 2...
>
> When writing your own assets, you may choose one or the other (or both) approaches...

> ## Part 1: Using the DuckDB Resource

Reviewer: ```suggestion
---

## Part 1: Using the DuckDB resource
```

Missing rule, and resource doesn't need to be capitalized

### Step 1: Configure the DuckDB resource

> To use the DuckDB resource, you add it to your `Definitions` object. The DuckDB resource requires some configuration. You must set a path to a duckdb database as the `database` configuration value. If the database does not already exist, it will be created for you:

Reviewer: ```suggestion
To use the DuckDB resource, you'll need to add it to your `Definitions` object. The DuckDB resource requires some configuration. You must set a path to a DuckDB database as the `database` configuration value. If the database does not already exist, it will be created for you:
```

```python file=/integrations/duckdb/tutorial/resource/configuration.py startafter=start_example endbefore=end_example
from dagster_duckdb import DuckDBResource

from dagster import Definitions

defs = Definitions(
    assets=[iris_dataset],
    resources={
        "duckdb": DuckDBResource(
            database="path/to/my_duckdb_database.duckdb",  # required
        )
    },
)

> # end_example

Reviewer: I think the start/end lines need to be adjusted for this

```

> ---

Reviewer: No rule needed here, since this is an H3.

def iris_dataset(duckdb: DuckDBResource) -> None:
    iris_df = pd.read_csv(
        "https://docs.dagster.io/assets/iris.csv",
        names=[
            "sepal_length_cm",
            "sepal_width_cm",
            "petal_length_cm",
            "petal_width_cm",
            "species",
        ],
    )

    with duckdb.get_connection() as conn:
        conn.execute("CREATE TABLE iris.iris_dataset AS SELECT * FROM iris_df")
```

In this example, you define an asset that fetches the Iris dataset as a Pandas DataFrame and renames the columns. Then, using the DuckDB resource, the DataFrame is stored in DuckDB as the `iris.iris_dataset` table.

</TabItem>

<TabItem name="Make existing tables available in Dagster">

#### Make existing tables available in Dagster

If you already have tables in DuckDB, you may want to have other assets in Dagster depend on those tables. You can accomplish this by creating [source assets](/concepts/assets/software-defined-assets#defining-external-asset-dependencies) for these tables.

```python file=/integrations/duckdb/tutorial/io_manager/source_asset.py
from dagster import SourceAsset

iris_harvest_data = SourceAsset(key="iris_harvest_data")
```

> In this example, you create a <PyObject object="SourceAsset" /> for a pre-existing table called `iris_harvest_data`.

Reviewer: ```suggestion
In this example, you're creating a <PyObject object="SourceAsset" /> for a pre-existing table called `iris_harvest_data`.
```

Nit - phrasing

</TabItem>

</TabGroup>

Now you can run `dagster dev` and materialize the `iris_dataset` asset from the Dagster UI.

> ---

Reviewer: Ugh I can't make a suggestion for this, but no rule is needed here since the following heading is an H3.

### Step 3: Define downstream assets

Once you have created an asset or source asset that represents a table in DuckDB, you will likely want to create additional assets that work with the data.

```python file=/integrations/duckdb/tutorial/resource/downstream.py startafter=start_example endbefore=end_example
from dagster import asset

# this example uses the iris_dataset asset from Step 1

@asset(deps=[iris_dataset])
def iris_setosa(duckdb: DuckDBResource) -> None:
    with duckdb.get_connection() as conn:
        conn.execute(
            "CREATE TABLE iris.iris_setosa AS SELECT * FROM iris.iris_dataset WHERE"
            " species = 'Iris-setosa'"
        )
```

> In this asset, you create second table that just contains the data for the Iris Setosa species. This asset has a dependency on the `iris_dataset` asset. To define this dependency, you provide the `iris_dataset` asset as a `dep` to the `iris_setosa` asset. You can then run the SQL query to create the table of Iris Setosa data.

Reviewer: ```suggestion
In this asset, you're creating a second table that contains only the data for the _Iris Setosa_ species. This asset has a dependency on the `iris_dataset` asset. To define this dependency, provide the `iris_dataset` asset as a `dep` to the `iris_setosa` asset. You can then run the SQL query to create the table of _Iris Setosa_ data.
```

Nit - some phrasing

I find the singular instance of `deps` kinda odd.

When finished, your code should look like the following:

```python file=/integrations/duckdb/tutorial/resource/full_example.py
import pandas as pd
from dagster_duckdb import DuckDBResource

from dagster import Definitions, SourceAsset, asset

iris_harvest_data = SourceAsset(key="iris_harvest_data")

@asset
def iris_dataset(duckdb: DuckDBResource) -> None:
    iris_df = pd.read_csv(
        "https://docs.dagster.io/assets/iris.csv",
        names=[
            "sepal_length_cm",
            "sepal_width_cm",
            "petal_length_cm",
            "petal_width_cm",
            "species",
        ],
    )

    with duckdb.get_connection() as conn:
        conn.execute("CREATE TABLE iris.iris_dataset AS SELECT * FROM iris_df")

@asset(deps=[iris_dataset])
def iris_setosa(duckdb: DuckDBResource) -> None:
    with duckdb.get_connection() as conn:
        conn.execute(
            "CREATE TABLE iris.iris_setosa AS SELECT * FROM iris.iris_dataset WHERE"
            " species = 'Iris-setosa'"
        )

defs = Definitions(
    assets=[iris_dataset],
    resources={
        "duckdb": DuckDBResource(
            database="path/to/my_duckdb_database.duckdb",
        )
    },
)
```

> ## Part 2: Using the DuckDB I/O Manager

Reviewer: ```suggestion
---

## Part 2: Using the DuckDB I/O manager
```

Missing rule, and `manager` doesn't need to be capitalized

You may want to use an I/O manager to handle storing DataFrames as tables in DuckDB and loading DuckDB tables as DataFrames in downstream assets. Using an I/O manager is not required, and you can reference [When to use I/O managers](TODO-LINK) to learn more.

This section of the guide focuses on storing and loading Pandas DataFrames in DuckDB, but Dagster also supports using PySpark and Polars DataFrames with DuckDB. The concepts from this guide apply to working with PySpark and Polars DataFrames, and you can learn more about setting up and using the DuckDB I/O manager with PySpark and Polars DataFrames in the [reference guide](/integrations/duckdb/reference).

### Step 1: Configure the DuckDB I/O manager

> The DuckDB I/O manager requires some configuration to connect to your database. You must provide a path where a DuckDB database will be created. Additionally, you can specify a `schema` where the DuckDB I/O manager will create tables.

Reviewer: It feels nitpicky, but this section doesn't talk about `Definitions` at all while the Step 1 of Part 1 does. If we're telling folks they can choose one approach or the other (or both), perhaps these similar sections should be similarly specific.

Author: good point

---

> This tutorial is divided in two parts. Both parts will create the same assets, but how the data is stored in DuckDb will differ. In [Part 1](#part-1-using-the-duckdb-resource) you will use the DuckDB resource to manually execute SQL queries to create tables and interact with those tables. In [Part 2](#part-2-using-the-duckdb-io-manager) you will transfer the responsibility for generating the SQL to create and interact with tables to the DuckDB I/O manager. When writing your own assets, you may choose one or the other (or both) approaches depending on your storage requirements. See [When to use I/O managers](TODO-LINK) to learn more.

Author: TODO - replace TODO link in this paragraph with the When to use IO managers section once that pr is merged

```python file=/integrations/duckdb/tutorial/resource/full_example.py
import pandas as pd
from dagster_duckdb import DuckDBResource

from dagster import Definitions, SourceAsset, asset

iris_harvest_data = SourceAsset(key="iris_harvest_data")

@asset
def iris_dataset(duckdb: DuckDBResource) -> None:
    iris_df = pd.read_csv(
        "https://docs.dagster.io/assets/iris.csv",
        names=[
            "sepal_length_cm",
            "sepal_width_cm",
            "petal_length_cm",
            "petal_width_cm",
            "species",
        ],
    )

    with duckdb.get_connection() as conn:
        conn.execute("CREATE TABLE iris.iris_dataset AS SELECT * FROM iris_df")

@asset(deps=[iris_dataset])
def iris_setosa(duckdb: DuckDBResource) -> None:
    with duckdb.get_connection() as conn:
        conn.execute(
            "CREATE TABLE iris.iris_setosa AS SELECT * FROM iris.iris_dataset WHERE"
            " species = 'Iris-setosa'"
        )

defs = Definitions(
    assets=[iris_dataset],
    resources={
        "duckdb": DuckDBResource(
            database="path/to/my_duckdb_database.duckdb",
        )
    },
)
```

## Part 2: Using the DuckDB I/O Manager

> You may want to use an I/O manager to handle storing DataFrames as tables in DuckDB and loading DuckDB tables as DataFrames in downstream assets. Using an I/O manager is not required, and you can reference [When to use I/O managers](TODO-LINK) to learn more.

Author: TODO - replace todo link with the when to use section when that pr merges

---

> ## Part 1: Using the DuckDB resource
> 
> ---
> 
> ### Step 1: Configure the DuckDB resource

Reviewer: ```suggestion
## Part 1: Using the DuckDB resource

### Step 1: Configure the DuckDB resource
```

Sorry if one of my suggestions did this - rules are only needed above H2 headings.

    )
```

> In this example, you define an asset that fetches the Iris dataset as a Pandas DataFrame, renames the columns, then returns the DataFrame. The type signature of the function tells the I/O manager what data type it is working with, so it is important to include the return type `pd.DataFrame`.

Reviewer: ```suggestion
In this example, you're defining an asset that fetches the Iris dataset as a Pandas DataFrame, renames the columns, then returns the DataFrame. The type signature of the function tells the I/O manager what data type it is working with, so it is important to include the return type `pd.DataFrame`.
```

iris_harvest_data = SourceAsset(key="iris_harvest_data")
```

> In this example, you create a <PyObject object="SourceAsset" /> for a pre-existing table containing iris harvests data. To make the data available to other Dagster assets, you need to tell the DuckDB I/O manager how to find the data.

Reviewer: ```suggestion
In this example, you're creating a <PyObject object="SourceAsset" /> for a pre-existing table containing iris harvests data. To make the data available to other Dagster assets, you need to tell the DuckDB I/O manager how to find the data.
```

Because you already supplied the database and schema in the I/O manager configuration in [Step 1: Configure the DuckDB I/O manager](#step-1-configure-the-duckdb-io-manager), you only need to provide the table name. This is done with the `key` parameter in `SourceAsset`. When the I/O manager needs to load the `iris_harvest_data` in a downstream asset, it will select the data in the `IRIS.IRIS_HARVEST_DATA` table as a Pandas DataFrame and provide it to the downstream asset.

</TabItem>
</TabGroup>

> ---

Reviewer: No rule is needed here

@asset
def iris_setosa(iris_dataset: pd.DataFrame) -> pd.DataFrame:
    return iris_dataset[iris_dataset["species"] == "Iris-setosa"]
```

> In this asset, you provide the `iris_dataset` asset as a dependency to `iris_setosa`. The `DuckDBPandasIOManager` will load this asset into memory as a Pandas DataFrame. Then you create a DataFrame that just contains the data for the Iris Setosa species and return the DataFrame. Then the `DuckDBPandasIOManager` will store the DataFrame as the `IRIS.IRIS_SETOSA` table in DuckDB.

Reviewer: ```suggestion
In this asset, you're providing the `iris_dataset` asset as a dependency to `iris_setosa`. The `DuckDBPandasIOManager` will load this asset into memory as a Pandas DataFrame. Next, a DataFrame that only contains the data for the _Iris Setosa_ species is created and returned. Then the `DuckDBPandasIOManager` will store the DataFrame as the `IRIS.IRIS_SETOSA` table in DuckDB.
```

@asset
def iris_setosa(iris_dataset: pd.DataFrame) -> pd.DataFrame:
    return iris_dataset[iris_dataset["species"] == "Iris-setosa"]
```

In this asset, you're providing the `iris_dataset` asset as a dependency to `iris_setosa`. By supplying `iris_dataset` as a parameter to `iris_setosa`, Dagster knows to use the `DuckDBPandasIOManager` to load this asset into memory as a Pandas DataFrame and pass it as an argument to `iris_setosa`. Next, a DataFrame that only contains the data for the _Iris Setosa_ species is created and returned. Then the `DuckDBPandasIOManager` will store the DataFrame as the `IRIS.IRIS_SETOSA` table in DuckDB.

> ---

Reviewer: Sorry I missed this last time - could you remove this?  The rule is only needed above H2 headings
